File: backend/src/models/phantom.py
"""
Pydantic models for Phantom data structures.

Based on patterns from congressional-trading-system schemas.py.
"""

from datetime import datetime
from enum import Enum
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class Position(str, Enum):
    """Investment position recommendation."""
    BULLISH = "bullish"
    BEARISH = "bearish"
    NEUTRAL = "neutral"
    AVOID = "avoid"


class Conviction(str, Enum):
    """Conviction level for a position."""
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


class TimestampMixin(BaseModel):
    """Mixin for timestamp fields."""
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


# Phantom Definition Models

class PhantomMemory(BaseModel):
    """A formative memory that shapes phantom judgment."""
    context: str = Field(..., description="Historical market situation")
    decision: str = Field(..., description="What the phantom did")
    reasoning: str = Field(..., description="Strategic logic behind the decision")
    outcome: str = Field(..., description="What happened as a result")
    lesson: str = Field(..., description="Strategic insight gained")


class PhantomDefinition(BaseModel):
    """Complete phantom persona definition."""
    investor_id: str = Field(..., description="Unique identifier (e.g., 'buffett')")
    name: str = Field(..., description="Full name (e.g., 'Warren Buffett')")
    era: str = Field(..., description="Investment era and context")
    philosophy: str = Field(..., description="Core investment philosophy")
    phantom_memories: List[PhantomMemory] = Field(default_factory=list)
    trigger_patterns: List[str] = Field(default_factory=list)
    blind_spots: List[str] = Field(default_factory=list)
    decision_framework: List[str] = Field(default_factory=list)


class PhantomSummary(BaseModel):
    """Brief phantom info for listing."""
    investor_id: str
    name: str
    philosophy: str


# Analysis Models

class AnalysisRequest(BaseModel):
    """Request for phantom analysis."""
    asset: str = Field(..., description="Asset to analyze (e.g., 'AAPL', 'BTC')")
    context: Optional[str] = Field(None, description="Additional market context")
    phantom_ids: Optional[List[str]] = Field(None, description="Specific phantoms to use (default: all)")


class PhantomAnalysis(BaseModel):
    """Analysis result from a single phantom."""
    phantom_id: str
    phantom_name: str
    position: Position
    conviction: Conviction
    reasoning: str = Field(..., description="Strategic reasoning for the position")
    key_factors: List[str] = Field(default_factory=list)
    risks: List[str] = Field(default_factory=list)
    blind_spots_acknowledged: List[str] = Field(default_factory=list)
    timestamp: datetime = Field(default_factory=datetime.utcnow)


class CouncilAnalysis(BaseModel):
    """Combined analysis from all phantoms (the council)."""
    asset: str
    analyses: List[PhantomAnalysis]
    consensus: Optional[Position] = Field(None, description="Consensus position if any")
    disagreements: List[str] = Field(default_factory=list, description="Key points of disagreement")
    timestamp: datetime = Field(default_factory=datetime.utcnow)


# Response Models

class PhantomListResponse(BaseModel):
    """Response for listing phantoms."""
    phantoms: List[PhantomSummary]
    total: int


class AnalysisResponse(BaseModel):
    """Response for a single phantom analysis."""
    success: bool
    analysis: Optional[PhantomAnalysis] = None
    error: Optional[str] = None


class CouncilResponse(BaseModel):
    """Response for council (all phantoms) analysis."""
    success: bool
    council: Optional[CouncilAnalysis] = None
    error: Optional[str] = None

File: frontend/src/types/phantom.ts
// Phantom types matching backend models

export type Position = "bullish" | "bearish" | "neutral" | "avoid";
export type Conviction = "high" | "medium" | "low";

export interface PhantomSummary {
  investor_id: string;
  name: string;
  philosophy: string;
}

export interface PhantomMemory {
  context: string;
  decision: string;
  reasoning: string;
  outcome: string;
  lesson: string;
}

export interface PhantomDefinition {
  investor_id: string;
  name: string;
  era: string;
  philosophy: string;
  phantom_memories: PhantomMemory[];
  trigger_patterns: string[];
  blind_spots: string[];
  decision_framework: string[];
}

export interface PhantomAnalysis {
  phantom_id: string;
  phantom_name: string;
  position: Position;
  conviction: Conviction;
  reasoning: string;
  key_factors: string[];
  risks: string[];
  blind_spots_acknowledged: string[];
  timestamp: string;
}

export interface CouncilAnalysis {
  asset: string;
  analyses: PhantomAnalysis[];
  consensus: Position | null;
  disagreements: string[];
  timestamp: string;
}

export interface PhantomListResponse {
  phantoms: PhantomSummary[];
  total: number;
}

export interface AnalysisResponse {
  success: boolean;
  analysis?: PhantomAnalysis;
  error?: string;
}

export interface CouncilResponse {
  success: boolean;
  council?: CouncilAnalysis;
  error?: string;
}

export interface AnalysisRequest {
  asset: string;
  context?: string;
  phantom_ids?: string[];
}

// Quick Scan types
export interface QuickScanRequest {
  symbols: string[];
  include_context?: boolean;
}

export interface OpportunityResult {
  symbol: string;
  score: number;
  consensus_position: string | null;
  consensus_strength: string | null;
  high_conviction_count: number;
  total_phantoms: number;
  key_insight: string;
  bullish_phantoms: string[];
  bearish_phantoms: string[];
  market_context: string | null;
  // Price data from Finnhub
  current_price: number | null;
  price_change: number | null;
  price_change_pct: number | null;
  day_high: number | null;
  day_low: number | null;
}

export interface QuickScanResponse {
  opportunities: OpportunityResult[];
  symbols_scanned: number;
  average_score: number;
}

// Opportunity Storage types
export interface SaveOpportunityRequest {
  symbol: string;
  scan_id?: string;
  opportunity_score: number;
  consensus_position?: string;
  consensus_strength?: string;
  high_conviction_count: number;
  total_phantoms: number;
  bullish_phantoms: string[];
  bearish_phantoms: string[];
  key_insight: string;
  market_context?: string;
  price_at_scan?: number;
}

export interface SaveOpportunitiesResponse {
  success: boolean;
  scan_id: string;
  saved_count: number;
  message: string;
}

export interface StoredOpportunity {
  id: string;
  symbol: string;
  scan_id?: string;
  scanned_at?: string;
  opportunity_score: number;
  consensus_position?: string;
  consensus_strength?: string;
  high_conviction_count: number;
  total_phantoms: number;
  bullish_phantoms: string[];
  bearish_phantoms: string[];
  key_insight: string;
  market_context?: string;
  price_at_scan?: number;
  current_price?: number;
  price_change_pct?: number;
  created_at?: string;
}
